File: app/src/main/java/com/example/mealrecipeapp/ui/home/RecipeAdapter.java
package com.example.mealrecipeapp.ui.home;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.example.mealrecipeapp.data.remote.response.Recipe;
import com.example.mealrecipeapp.databinding.RecipeItemBinding;

import java.util.List;

public class RecipeAdapter extends RecyclerView.Adapter<RecipeAdapter.RecipeViewHolder> {

    private final OnClickListener onClickListener;
    private List<Recipe> recipes;

    public RecipeAdapter(OnClickListener onClickListener) {
        this.onClickListener = onClickListener;
    }

    public void setRecipes(List<Recipe> recipes) {
        this.recipes = recipes;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public RecipeViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        RecipeItemBinding view = RecipeItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new RecipeViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull RecipeViewHolder holder, int position) {
        Recipe recipe = recipes.get(position);
        holder.bind(recipe, onClickListener);
    }

    @Override
    public int getItemCount() {
        return (recipes != null) ? recipes.size() : 0;
    }

    public static class RecipeViewHolder extends RecyclerView.ViewHolder {
        private final RecipeItemBinding binding;

        public RecipeViewHolder(RecipeItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(Recipe recipe, OnClickListener onClickListener) {
            binding.titileTextView.setText(recipe.getTitle());
            binding.voteTextView.setText(recipe.getAggregateLikes() + " likes");
            binding.timeTextView.setText(recipe.getReadyInMinutes() + " min");
            binding.servingTextView.setText(recipe.getServings() + " servings");
            Glide.with(itemView)
                    .load(recipe.getImage())
                    .into(binding.posterImageView);

            binding.getRoot().setOnClickListener(v -> onClickListener.onItemClick(recipe));
        }
    }
    // Interface for handling item clicks
    public interface OnClickListener {
        void onItemClick(Recipe recipe);
    }
}

File: app/src/main/java/com/example/mealrecipeapp/ui/home/RecipeAdapter.kt
package com.example.mealrecipeapp.ui.home

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.mealrecipeapp.R
import com.example.mealrecipeapp.data.remote.response.Recipe
import com.example.mealrecipeapp.databinding.RecipeItemBinding

class RecipeAdapter(
    private val clickListener: (Long) -> Unit
) : RecyclerView.Adapter<RecipeAdapter.RecipeViewHolder>() {
    private var recipes: List<Recipe> = listOf()
    fun setRecipes(recipes: List<Recipe>) {
        this.recipes = recipes
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecipeViewHolder {
        val view = RecipeItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return RecipeViewHolder(view)
    }

    override fun onBindViewHolder(holder: RecipeViewHolder, position: Int) {
        val recipe = recipes[position]
        holder.bind(recipe, clickListener)
    }

    override fun getItemCount(): Int {
        return recipes.size
    }

    class RecipeViewHolder(private val binding: RecipeItemBinding) : RecyclerView.ViewHolder(
        binding.root
    ) {
        fun bind(recipe: Recipe, clickListener: (Long) -> Unit) {
            binding.titileTextView.text = recipe.title
            binding.voteTextView.text = itemView.resources.getString(R.string.like_counts, recipe.aggregateLikes.toString())
            binding.timeTextView.text = itemView.resources.getString(R.string.time_counts, recipe.readyInMinutes.toString())
            binding.servingTextView.text = itemView.resources.getString(R.string.serving_counts, recipe.servings.toString())
            Glide.with(itemView)
                .load(recipe.image)
                .into(binding.posterImageView)
            binding.root.setOnClickListener { clickListener(recipe.id) }
        }
    }
}
